engines/sherlock/fonts.cpp

```diff
@@ -372,28 +398,165 @@ int Fonts::stringHeight(const Common::String &str) {
 }
 
+Common::Array<Common::String> Fonts::wordWrap(const Common::String &str, uint maxWidth, Common::String &rem, uint maxChars, uint maxLines, bool skipHeadAt) {
+	Common::Array<Common::String> lines;
+	int strIdx = 0;
+
+	bool isInEucEscape = false;
+	do {
+		uint width = 0;
+		uint numChars = 0;
+		int spaceIdx = 0;
+		bool spacePNeedsEndEscape = false;
+		bool spacePNeedsBeginEscape = false;
+		int lineStartIdx = strIdx;
+		// Invariant: lastCharIdx is either -1
+		// or is exactly one character behind strP
+		// and in the same escape state.
+		int lastCharIdx = -1;
+		bool isLineStartPInEucEscape = isInEucEscape;
+
+		if (skipHeadAt) {
+			// If the first character is a '@' flagging a title line, then move
+			// past it, so the @ won't be included in the line width calculation
+			if (strIdx + 1 < (int)str.size() && str[strIdx] == '@' && str[strIdx + 1] != '$')
+				++strIdx;
+		}
+
+		// Find how many characters will fit on the next line
+		while (width < maxWidth && numChars < maxChars && strIdx < (int)str.size() && str[strIdx] != '\n') {
+			if (_isModifiedEucCn) {
+				byte curChar = str[strIdx];
+				byte nextChar = strIdx + 1 < (int)str.size() ? str[strIdx + 1] : 0;
+				if (!isInEucEscape && curChar == '@' && nextChar == '$') {
+					width += charWidth(' ');
+					numChars++;
+					if (lineStartIdx != strIdx) {
+						spaceIdx = strIdx;
+						spacePNeedsEndEscape = isInEucEscape;
+						spacePNeedsBeginEscape = true;
+					}
+					lastCharIdx = -1;
+					strIdx += 2;
+					isInEucEscape = true;
+					continue;
+				}
+
+				if (isInEucEscape && curChar == '$' && nextChar == '@') {
+					width += charWidth(' ');
+					numChars++;
+					spaceIdx = strIdx;
+					lastCharIdx = -1;
+					strIdx += 2;
+					spacePNeedsEndEscape = isInEucEscape;
+					spacePNeedsBeginEscape = false;
+					isInEucEscape = false;
+					continue;
+				}
+
+				if (curChar >= 0x41 && nextChar >= 0x41 && (isInEucEscape || ((curChar >= 0xa1) && (nextChar >= 0xa1)))) {
+					width += kChineseWidth;
+					lastCharIdx = strIdx;
+					strIdx += 2;
+					numChars++;
+					continue;
+				}
+			}
+
+			// Keep track of the last space
+			if (str[strIdx] == ' ') {
+				spacePNeedsEndEscape = isInEucEscape;
+				spacePNeedsBeginEscape = isInEucEscape;
+				spaceIdx = strIdx;
+			}
+			lastCharIdx = strIdx;
+			width += charWidth(str.c_str(), strIdx);
+			numChars++;
+		}
+
+		bool previousEucEscape = isInEucEscape;
+
+		// If the line was too wide to fit on a single line, go back to the last space
+		// if there was one, or otherwise simply break the line at this point
+		if (width >= maxWidth || numChars >= maxChars) {
+			if (spaceIdx > 0) {
+				previousEucEscape = spacePNeedsEndEscape;
+				isInEucEscape = spacePNeedsBeginEscape;
+				strIdx = spaceIdx;
+			} else if (lastCharIdx > 0 && lastCharIdx != lineStartIdx) {
+				strIdx = lastCharIdx;
+			}
+		}
+
+		Common::String line = str.substr(lineStartIdx, strIdx - lineStartIdx);
+		assert(!line.contains('\n'));
```

Comment:
Is it always the case, so we `assert()`? Or do we want explicitly catch such cases?

Reply:
There are two cases where this triggers:
1. Bug in wrapping that results in line containing \n
2. Invalid input containing \n as a part of multibyte sequence
Either case will cause an assert failure when trying to actually write the string. It's better to catch it earlier